crates/sui-indexer/src/store/pg_partition_manager.rs

```diff
@@ -0,0 +1,139 @@
+         JOIN pg_class child ON pg_inherits.inhrelid = child.oid
+         JOIN pg_namespace nmsp_parent ON nmsp_parent.oid = parent.relnamespace
+         JOIN pg_namespace nmsp_child ON nmsp_child.oid = child.relnamespace
+WHERE parent.relkind = 'p'
+GROUP BY table_name;
+"#;
+
+#[derive(Clone)]
+pub struct PgPartitionManager {
+    cp: PgConnectionPool,
+}
+
+#[derive(Clone)]
+pub struct EpochPartitionData {
+    last_epoch: u64,
+    next_epoch: u64,
+    last_epoch_start_cp: u64,
+    next_epoch_start_cp: u64,
+}
+
+impl From<EpochToCommit> for EpochPartitionData {
+    fn from(epoch: EpochToCommit) -> Self {
+        let last_epoch = epoch.last_epoch.as_ref().map(|e| e.epoch).unwrap_or(0);
+        let next_epoch = epoch.new_epoch.epoch;
+        let last_epoch_start_cp = epoch.last_epoch.map(|e| e.first_checkpoint_id).unwrap_or(0);
+        let next_epoch_start_cp = epoch.new_epoch.first_checkpoint_id;
+        Self {
+            last_epoch,
+            next_epoch,
+            last_epoch_start_cp,
+            next_epoch_start_cp,
+        }
+    }
+}
+
+impl PgPartitionManager {
+    pub fn new(cp: PgConnectionPool) -> Result<Self, IndexerError> {
+        let manager = Self { cp };
+        let tables = manager.get_table_partitions()?;
+        info!(
+            "Found {} tables with partitions : [{:?}]",
+            tables.len(),
+            tables
+        );
+        Ok(manager)
+    }
+
+    pub fn get_table_partitions(&self) -> Result<BTreeMap<String, u64>, IndexerError> {
+        #[derive(QueryableByName, Debug, Clone)]
+        struct PartitionedTable {
+            #[diesel(sql_type = VarChar)]
+            table_name: String,
+            #[diesel(sql_type = VarChar)]
+            last_partition: String,
+        }
+
+        Ok(
+            read_only_blocking!(&self.cp, |conn| diesel::RunQueryDsl::load(
+                diesel::sql_query(GET_PARTITION_SQL),
+                conn
+            ))?
+            .into_iter()
+            .map(|table: PartitionedTable| {
+                u64::from_str(&table.last_partition)
+                    .map(|last_partition| (table.table_name, last_partition))
+                    .map_err(|e| anyhow!(e))
+            })
+            .collect::<Result<_, _>>()?,
+        )
+    }
+
+    pub fn advance_table_epoch_partition(
+        &self,
+        table: &str,
+        data: &EpochPartitionData,
+    ) -> Result<(), IndexerError> {
+        let (last_epoch, next_epoch, last_epoch_start_cp, next_epoch_start_cp) = (
+            data.last_epoch,
+            data.next_epoch,
+            data.last_epoch_start_cp,
+            data.next_epoch_start_cp,
+        );
+        // first partition has been created in /migrations
+        if next_epoch == 0 {
+            tracing::warn!("skipping epoch 0");
+            return Ok(());
+        }
+        let detach_last_partition =
+            format!("ALTER TABLE {table} DETACH PARTITION {table}_partition_{last_epoch};");
```

Comment:
if my understanding's correct, we do not want to detach a partition until it's offloaded to remote storage and we can delete it? Why would we want to detach it rather than leaving it as is before pruning?

Reply:
this is not really to detach the partition, but to update the ending checkpoint of this partition, the way to do this is by 
- detaching a partition with max_value as the ending checkpoint
- move around data and re-attaching the partition with last_cp of the current epoch, as the endpoint checkpoint of the partition.